lib/kase/switcher.rb

```diff
@@ -34,39 +34,30 @@ def validate!
     end
 
     def switch(&block)
-      context = eval("self", block.binding)
-      dsl = DSL.new(self, context)
-      dsl.__call(&block)
+      DSL.call(self, &block)
       validate!
       result
     end
 
-    class DSL
-      def initialize(switcher, context)
-        @__switcher = switcher
-        @__context = context
-      end
+    module DSL
+      module_function
 
-      def __call(&block)
-        @__context.instance_variables.each do |name|
-          next if name.to_s =~ /^@__/
-          instance_variable_set(name, @__context.instance_variable_get(name))
-        end
+      def call(switcher, &block)
+        context = eval("self", block.binding)
 
-        instance_eval(&block)
+        # Preserve original on method
+        original_on = context.method(:on) if defined? context.on
 
-        instance_variables.each do |name|
-          next if name.to_s =~ /^@__/
-          @__context.instance_variable_set(name, instance_variable_get(name))
+        # Define new on method
+        context.define_singleton_method(:on) do |*pattern, &block|
```

Comment:
This is probably OK as long `#on` won't overwrite an existing `#on`. I don't know of many instances where that might be. In the code base I am working on now, we have experimented with an event manager that might use `on`. That might even be ok since the `on` we want inside a kase block is different the an `on` we want outside of it. However, in that case, probably want `on` to get restored.